### PR TITLE
update refactor & fix update sequence via binary

### LINKED ISSUE #516: `eris update` on binaries logic not currently working
**Problem** -- when updating from a binary installation, `eris update` still tries to install via go. The [current code](https://github.com/eris-ltd/eris-cli/blob/v0.11.1/util/update_tool.go#L96-L102) which controls a binary install will pull the binary OK, but it will still try to check the go path (there is no return in that `if` statement, so the func will just continue down to the "install via go" [section here](https://github.com/eris-ltd/eris-cli/blob/v0.11.1/util/update_tool.go#L117-L126).

**Solution** -- refactor the if statement such that it: downloads binary to proper location; runs migrate; runs `version()` and notifies user of successful installation; then it should exit rather than continue with the `UpdateEris()` function as currently structured.

### PR DESCRIPTION
**Pull request by zramsay**
- This PR refactors the `eris update` command and fixes its inability to update from binary installation (rather than go). Closes #516
- A (y/n) prompt was added to confirm the appropriate path/binary is being updated (go or binary)
- `eris update` via go has a `--branch=someBranch` feature that now works even if `eris` was installed via binary. This is accomplished by using a disposable build container (`perform.DockerBuild()`) and copy-ing out the binary (`data.ExportData()`). Without specifying a branch, `eris update` will default to master, thus grabbing the latest release.

### CODE REVIEW

**Comment by zramsay** [comment]
blocking on #586 which will simplify part of the process

**Comment by zramsay** [comment]
this works with this manual test: (assuming the test-mach has been created)
```
go install ./cmd/eris
docker-machine scp $(which eris) test-mach:/usr/bin
docker-machine ssh test-mach
eris update --branch=fake-version
eris version
```
final `eris version` will output 2.9.9 from that fake branch

File: data/operate.go [hunk 1]
```diff
@@ -160,9 +160,9 @@ func ExportData(do *definitions.Do) error {
 		reader, writer := io.Pipe()
 		defer reader.Close()
 
-		if err := checkErisContainerRoot(do, "export"); err != nil {
-			return err
-		}
+		//if err := checkErisContainerRoot(do, "export"); err != nil {
+		//	return err
+		//}
```
Comment (zramsay) [inline data/operate.go]: need smarter solution for this...

File: data/operate.go [hunk 2]
```diff
@@ -238,6 +238,7 @@ func MoveOutOfDirAndRmDir(src, dest string) error {
 // check path for ErisContainerRoot
 // XXX this is opiniated & we may want to change in future
 // for more flexibility with filesystem of data conts
+// [zr] yes, it is opiniated, silenced in ExportData for flexibility
```
Comment (zramsay) [inline data/operate.go]: see above comment; commenting out that functionality was req'd to use `ExportData()` to cp the binary out

File: perform/perform.go [hunk 3]
```diff
@@ -640,13 +642,66 @@ func DockerRemove(srv *def.Service, ops *def.Operation, withData, volumes, force
...
+		defer close(ch)
+
+		if err := util.DockerClient.BuildImage(imgOpts); err != nil {
+			ch <- err
+		}
+	}()
+	jsonmessage.DisplayJSONMessagesStream(r, writer, os.Stdout.Fd(), term.IsTerminal(os.Stdout.Fd()), nil)
+	if err, ok := <-ch; ok {
+		// XXX err is not thrown!
+		return err
```
Comment (zramsay) [inline perform/perform.go]: a failed build (e.g. branch specified does not exist) does not return an error.

File: update/binary.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+		}
+
+	} else {
+		cpString := fmt.Sprintf("%s %s", newPath, oldPath)
+		log.Warn(`
+To complete the update on windows, run:
+
+rm ` + oldPath + `
+mv ` + cpString + `
+chmod +x ` + oldPath + `
```
Comment (zramsay) [inline update/binary.go]: needs to be tested properly on windows

File: update/binary.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+	return nil
+}
+
+
+
+// XXX code below is ~ replaced by code above. 
+// left here for legacy reasons / is a complementary
+// approach to the build container that we may want
+// to consider ... ?
+/*func DownloadLatestBinaryRelease(binPath string) (string, error) {
```
Comment (zramsay) [inline update/binary.go]: not sure what to do with all this code...

File: update/update.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+		if err := UpdateErisGo(do); err != nil {
+			return err
+		}
+	} else if whichEris == "binary" {
+		if err := UpdateErisViaBinary(do.Branch, binPath); err != nil {
+			return err
+		}
+
+		// XXX deprecate ... ?
+		//if err := UpdateErisBinaryOLD(binPath); err != nil {
```
Comment (zramsay) [inline update/update.go]: see previous comment

File: update/update.go [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+	log.WithField("branch", branch).Warn("Building eris binary in container with:")
+	if err := BuildErisBinContainer(branch, binPath); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// TODO => deprecate ... ?
```
Comment (zramsay) [inline update/update.go]: ibid

File: update/update.go [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+	return nil
+}
+
+func UpdateErisViaBinary(branch, binPath string) error {
+
+	log.WithField("branch", branch).Warn("Building eris binary in container with:")
+	if err := BuildErisBinContainer(branch, binPath); err != nil {
+		return err
+	}
+
```
Comment (zramsay) [inline update/update.go]: log.Warn("The marmots have updated...")

File: update/update.go [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+	log.WithField("=>", ver).Warn("The marmots have updated Eris successfully")
+	return nil
+}
+
+// returns a string of each the type of installation
+// and path to the binary; the latter is only used for
+// non-go binary installation
+func GoOrBinary() (string, string, error) {
```
Comment (zramsay) [inline update/update.go]: this function feels a bit messy

File: update/binary.go [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+	dockerfile := `FROM ` + baseImage + `
+ENV NAME         eris-cli
+ENV REPO 	 eris-ltd/$NAME
+ENV BRANCH       ` + branch + `
+ENV CLONE_PATH   $GOPATH/src/github.com/$REPO
+ENV GO15VENDOREXPERIMENT 1
+
+RUN mkdir --parents $CLONE_PATH
+
+RUN git clone --quiet https://github.com/$REPO $CLONE_PATH
```
Comment (pietv) [inline update/binary.go]: Love the idea of building an ad-hoc image to prepare the binary inside a container. Neat! 👍

File: update/binary.go (hunk 4 above)
Comment (pietv) [inline update/binary.go]: Great solution for just suggesting to replace the binary on Windows! No need to support a dedicated updater! :)

Comment (pietv) [inline update/binary.go]: Better `del /f`, `ren` and no `chmod` on Windows.

File: update/binary.go (hunk 5 above)
Comment (pietv) [inline update/binary.go]: I'd drop it.

File: perform/perform.go (hunk 3 above)
Comment (pietv) [inline perform/perform.go]: :+1:

File: update/update.go (hunk 9 above)
Comment (pietv) [inline update/update.go]: Looks good to me.

File: update/update.go [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+	which, err := exec.Command("which", "eris").CombinedOutput()
+	if err != nil {
+		return "", "", err
+	}
+
```
Comment (pietv) [inline update/update.go]: Can you do away without invoking `which`? It's not present by default on Windows, so you can't expect it's there. Also, on Cygwin `which` and `exec.LookPath` don't produce the same results (even if installed binary is there)?

File: update/update.go [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+			binWarn := fmt.Sprintf(`The marmots have detected a installation located at: (%s)
+Continuing will update eris to latest master branch or the one your've specified.
+Do you wish to continue?`, trimWhich)
+			if util.QueryYesOrNo(binWarn) == util.Yes {
+				return "binary", erisLook, nil
+			} else {
+				return "", "", fmt.Errorf("Permission to update denied; exiting.")
+			}
+		}
+	} else {
```
Comment (pietv) [inline update/update.go]: Duplication. :)

Also, maybe reword this sentence `The marmots have detected a go installation located at: (/Users/peter/bin/eris)` to `... detected a source installation`, because the phrase a bit confusing, sounds like Go itself is found at that path.

**Comment by pietv** [comment]
LGTM :+1: Super job!

File: update/update.go (hunk 11 above)
Comment (zramsay) [inline update/update.go]: done.
